File: second_sino_japanese_war/submissions/accepted/alexis_bfs.cpp
/*
 * Solution en O(m*p) avec un bfs
 * @EXPECTED_RESULTS@: CORRECT
 */

#include<bits/stdc++.h>

using namespace std;

const int INF = numeric_limits<int>::max()/2;

int girth(vector<vector<int>>& adj, int source) {
    vector<int> d(adj.size(), INF);
    d[source] = 0;
    queue<int> q;
    q.push(source);

    int shortest_cycle = INF;
    vector<int> p(adj.size());

    while(!q.empty()){
        int current = q.front(); q.pop();
        for(auto& v : adj[current]) {
            if(v == p[current]) continue;
            if(current == v) continue;

            if(d[v] == INF){
                p[v] = current;
                d[v] = d[current] + 1;
                q.push(v);
            }else {
                shortest_cycle = min(shortest_cycle, d[current] + d[v] + 1);
                continue;
            }
        }
    }

    return shortest_cycle;
}

void solve() {
    int n, m, p; cin >> n >> m >> p;
    vector<vector<int>> adj(n);
    vector<int> candidates(m);
    for(auto& i : candidates) cin >> i;

    for(int i = 0; i < p; i++) {
        int u, v; cin >> u >> v;
        adj[u].push_back(v);
        adj[v].push_back(u);
    }

    int ans = INF;
    int best_idx = -1;
    //to remove
    unordered_map<int, vector<int>> all_ans;
    for(auto& c : candidates) {
        int g =  girth(adj, c);
        all_ans[g].push_back(c);
        if (g < ans) {
            ans = g;
            best_idx = c;
        }
    }

    cout << all_ans[ans].size() << endl;
    for(auto& k : all_ans[ans]){
        cout << k << " ";
    }
    cout << endl;

}

signed main() {
    cin.tie(0);
    ios_base::sync_with_stdio(false);
    solve();
    return 0;
}

File: second_sino_japanese_war/submissions/accepted/alexis_bfs.py
#
# Solution en O(m*p) avec un bfs
# @EXPECTED_RESULTS@: CORRECT
#

import collections

INF = 1e18+1

def girth(adj, source):
    distance = [INF for i in range(len(adj))]
    distance[source] = 0

    parent = [-1 for i in range(len(adj))]

    ans = INF

    queue = []
    queue.append(source)
    while(len(queue) > 0):
        current = queue.pop(0)
        for v in adj[current]:
            if distance[v] == INF:
                distance[v] = distance[current] + 1
                parent[v] = current
                queue.append(v)
            elif v != parent[current] and current != v:
                ans = min(ans, distance[current] + distance[v] + 1)
    return ans


def solve():
    n, m, p = list(map(int,input().split(" ")))

    candidates = list(map(int, input().split(" ")))

    adj = [[] for _ in range(n)]

    for _ in range(p):
        u, v = list(map(int, input().split(" ")))
        adj[u].append(v)
        adj[v].append(u)

    best_girth = INF

    anss = collections.defaultdict(list)
    for c in candidates:
        g = girth(adj, c)
        anss[g].append(c)
        if g < best_girth:
            best_girth = g

    print(len(anss[best_girth]))
    print(" ".join(list(map(str,anss[best_girth]))))

solve()